### PR TITLE
`<deque>`: Properly destroy (fancy) pointers to blocks in the internal map

### LINKED ISSUE #2769: <deque>: For allocators where allocator_traits<T>::pointer is an object, destructors aren't always called
When using a custom allocator that defines `allocator_traits<T>::pointer` as a class or struct, `std::deque` constructs more instances of that `pointer` object than it destroys. So if `pointer` has non-trivial constructors or destructors, some destructors will not be called.

```
> cl repro.cpp /std:c++latest

#include <type_traits>
#include <cstdio>
#include <cstdlib>
#include <compare>
#include <iterator>
#include <deque>

int numPtrs = 0;

template<class T>
class FancyPointer
{
    T* p{};

    template<class U>
    friend class FancyPointer;

    template<class U> requires std::is_convertible_v<U*, T*>
    FancyPointer(U* p)
        : FancyPointer()
    {
        this->p = p;
    }
    
    template<class U>
    friend class FancyAllocator;
public:
    
    using value_type = T;
    using pointer = FancyPointer<T>;
    using difference_type = ptrdiff_t;
    using iterator_category = std::random_access_iterator_tag;

    template<class U = T> requires (!std::is_void_v<U>)
    using reference = U&;

    FancyPointer() {
        ++numPtrs;
    }

    ~FancyPointer() {
        --numPtrs;
    }

    FancyPointer(const FancyPointer& fp)
        : FancyPointer()
    {
        p = fp.p;
    }

    FancyPointer(FancyPointer&& fp)
        : FancyPointer()
    {
        p = fp.p;
        fp.p = nullptr;
    }

    template<class U> requires std::is_convertible_v<U*, T*>
    FancyPointer(const FancyPointer<U>& fp)
        : FancyPointer()
    {
        p = fp.p;
    }

    FancyPointer(std::nullptr_t)
        : FancyPointer()
    {
        p = nullptr;
    }

    FancyPointer& operator=(const FancyPointer& fp) {
        p = fp.p;
        return *this;
    }

    template<class U> requires std::is_convertible_v<U*, T*>
    FancyPointer& operator=(const FancyPointer<U>& fp) {
        p = fp.p;
        return *this;
    }
    
    FancyPointer& operator=(FancyPointer&& fp) {
        p = fp.p;
        fp.p = nullptr;
        return *this;
    }

    FancyPointer& operator=(std::nullptr_t) {
        reset();
        return *this;
    }

    T* get() const {
        return p;
    }

    void reset() {
        p = nullptr;
    }

    explicit operator bool() const {
        return p != nullptr;
    }

    T* operator->() const {
        return p;
    }

    template<class U = T> requires (!std::is_void_v<U>)
    U& operator*() const {
        return *p;
    }

    template<class U = T> requires (!std::is_void_v<U>)
    U& operator[](ptrdiff_t index) const {
        return p[index];
    }

    FancyPointer& operator++() {
        ++p;
        return *this;
    }

    FancyPointer operator++(int) {
        auto tmp = *this;
        ++p;
        return tmp;
    }

    FancyPointer& operator--() {
        --p;
        return *this;
    }

    FancyPointer operator--(int) {
        auto tmp = *this;
        --p;
        return tmp;
    }

    FancyPointer& operator+=(ptrdiff_t off) {
        p += off;
        return *this;
    }

    FancyPointer& operator-=(ptrdiff_t off) {
        p -= off;
        return *this;
    }

    FancyPointer operator+(ptrdiff_t off) const {
        return { p + off };
    }

    FancyPointer operator-(ptrdiff_t off) const {
        return { p - off };
    }

    ptrdiff_t operator-(const FancyPointer& fp) const {
        return { p - fp.p };
    }

    template<class U> requires (!std::is_void_v<U> && std::is_same_v<U, T>)
    static FancyPointer<U> pointer_to(U& u) {
        return { &u };
    }

    template<class T2>
    bool operator==(const FancyPointer<T2>& right) const {
        return get() == right.get();
    }

    bool operator==(std::nullptr_t) const {
        return get() == nullptr;
    }

    template<class T2>
    std::strong_ordering operator<=>(const FancyPointer<T2>& right) {
        return get() <=> right.get();
    }

    std::strong_ordering operator<=>(std::nullptr_t) {
        return get() <=> nullptr;
    }
};

template <class T>
class FancyAllocator
{
public:
    using value_type = T;
    using pointer = FancyPointer<T>;
    using const_pointer = FancyPointer<const T>;
    using difference_type = ptrdiff_t;
    using size_type = std::size_t;

    template <class U>
    struct rebind {
        using other = FancyAllocator<U>;
    };

    FancyAllocator() {}
    ~FancyAllocator() {}

    template<class U>
    FancyAllocator(const FancyAllocator<U>& that) {}

    pointer allocate(size_type n) {
        void* p = ::operator new(n * sizeof(T), std::align_val_t(alignof(T)));
        return pointer{ static_cast<T*>(p) };
    }

    void deallocate(pointer fp, size_type n) {
        ::operator delete(fp.p, std::align_val_t(alignof(T)));
    }

    FancyAllocator select_on_container_copy_construction() const {
        return *this;
    }

    using propagate_on_container_copy_assignment = std::false_type;
    using propagate_on_container_move_assignment = std::true_type;
    using propagate_on_container_swap = std::true_type;
    using is_always_equal = std::true_type;
};

template <class T1, class T2>
inline bool operator==(const FancyAllocator<T1>& a, const FancyAllocator<T2>& b) noexcept {
    return true;
}

template <class T1, class T2>
inline bool operator!=(const FancyAllocator<T1>& a, const FancyAllocator<T2>& b) noexcept {
    return false;
}


int main(int argc, char* argv[])
{
    {
        std::deque<int, FancyAllocator<int>> stuff;
        stuff.push_back(1);
    }

    printf("FancyPointers (should be zero) : %d\n", numPtrs);
    return 0;
}
```

**Results**

`FancyPointer` has a counter in the constructor and destructor to count the number of live instances. After the `deque` has been destroyed, the counter should be zero, showing that all `FancyPointer` destructors have run correctly. However, the program outputs '7'. If `std::deque` in the above program is replaced with `std::list` or `std::vector`, the program will output '0' as expected.

**STL version**
Microsoft Visual Studio Enterprise 2022 (64-bit) - Preview
Version 17.3.0 Preview 1.1

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
The internal map array is fully constructed, so it should also be fully destroyed before deallocation.

Fixes #2769.

### CODE REVIEW

File: stl/inc/deque [hunk 1]
```diff
@@ -1466,7 +1466,7 @@ private:
             _Uninitialized_value_construct_n_unchecked1(_Myptr, _Count); // clear rest to initial block
         }
 
-        _Destroy_range(_Map() + _Myboff, _Map() + _Mapsize());
+        _Destroy_range(_Map(), _Map() + _Mapsize());
```
Comment (miscco) [inline stl/inc/deque:1469]: I find it super dubious that we always destroy the objects in the range but only sometimes deallocate the range.

That said the whole implementation is a bit special so ....

File: tests/std/tests/GH_002769_handle_deque_block_pointers/test.cpp [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+#include <cassert>
+#include <cstddef>
+#include <deque>
+#include <iterator>
+#include <memory>
+#include <type_traits>
+
+using namespace std;
+
+size_t fancy_counter{};
```
Comment (miscco) [inline tests/std/tests/GH_002769_handle_deque_block_pointers/test.cpp:13]: I would prefer to use 0 here 
```suggestion
size_t fancy_counter = 0;
```

File: tests/std/tests/GH_002769_handle_deque_block_pointers/test.cpp [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+template <class T>
```
Comment (miscco) [inline tests/std/tests/GH_002769_handle_deque_block_pointers/test.cpp:15]: That is some pretty extensive work needed for a small test. 

I am wondering whether it would make sense to reuse that machinery and test all containers for internal memory leaks?

File: stl/inc/deque (hunk 1 above)
Comment (frederick-vs-ja) [reply stl/inc/deque:1469]: I think the range is deallocate if and only if `_Map()` is not null. And when `_Map()` is null, the `_Destroy_range` call is no-op.

Comment (miscco) [reply stl/inc/deque:1469]: Yes, but then why not pull it into the if?

Comment (frederick-vs-ja) [reply stl/inc/deque]: Nice. I'm doing this now.

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

**Review by CaseyCarter — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]

File: stl/inc/deque [hunk 4]
```diff
@@ -693,7 +693,7 @@ private:
 
     void _Construct_n(size_type _Count, const _Ty& _Val) { // construct from _Count * _Val
         _Tidy_guard<deque> _Guard{this};
-        for (; 0 < _Count; --_Count) {
+        for (; _Count > 0; --_Count) {
```
Comment (StephanTLavavej) [inline stl/inc/deque:696]: No change requested: We generally try to avoid mixing unrelated cleanups into behavioral changes (as this increases the difficulty of reviewing and understanding source history). This isn't an ironclad rule - we're human and it's easier to fold changes into an existing PR, so a small amount of cleanup can be fine. In this case I think it's fine, but I wanted to mention it.

File: tests/std/tests/GH_002769_handle_deque_block_pointers/test.cpp [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    }
+
+    friend bool operator==(const counting_ptr& p, nullptr_t) noexcept {
+        return p.p_ == nullptr;
+    }
+
+#if _HAS_CXX20
+    friend bool operator==(const counting_ptr& lhs, const counting_ptr& rhs) = default;
+
+    friend auto operator<=>(const counting_ptr& lhs, const counting_ptr& rhs) = default;
```
Comment (StephanTLavavej) [inline tests/std/tests/GH_002769_handle_deque_block_pointers/test.cpp:142]: No change requested: In product code, we conventionally don't name parameters of defaulted functions. On the other hand, in test code we're somewhat less rigorous about conventions. On the third hand, `operator=(const counting_ptr&)` didn't name its parameter, which is an argument for being locally consistent here. On the fourth and final hand, this isn't worth resetting testing. :joy_cat:

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
I've resolved a trivial adjacent-add merge conflict in `tests/std/test.lst`.

**Comment by StephanTLavavej** [comment]
Thanks for fixing this correctness bug! :lady_beetle: :tada: :smile_cat:
